### PR TITLE
Remove readFrom from org.elasticsearch.search

### PR DESCRIPTION
**Pull request by nik9000**
Replace with a constructor that takes StreamInput or a static method.

In one case (ValuesSourceType) we no longer need to serialize the data
at all!

Relates to #17085

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/profile/CollectorResult.java [hunk 1]
```diff
@@ -148,9 +148,4 @@ public void writeTo(StreamOutput out) throws IOException {
             child.writeTo(out);
         }
     }
-
-    @Override
-    public Object readFrom(StreamInput in) throws IOException {
-        return new CollectorResult(in);
-    }
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/profile/CollectorResult.java]: nit: should we move down the corresponding ctor like we do in most of the other cases?

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java [hunk 2]
```diff
@@ -19,34 +19,9 @@
...
-
-import java.io.IOException;
-
-/*
- * The ordinal values for this class are tested in ValuesSourceTypeTests to
- * ensure that the ordinal for each value does not change and break bwc
- */
-public enum ValuesSourceType implements Writeable<ValuesSourceType> {
-
+public enum ValuesSourceType  {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: just out of curiosity, why is this enum not serialized any more?

**Comment by cbuescher** [comment]
@nik9000 one nit, one question, otherwise looks good.

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: When I refactored serialization of aggregations it was a parameter we were serializing but it was inconsistent to do so - every single aggregation subclass knew the value of the parameter. So I made it a required argument to the reading ctor. I kept expecting to hit an aggregation that needed to serialize it but I never did.

**Comment by nik9000** [comment]
Thanks for reviewing @cbuescher !

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java]: thanks, that explains it.

**Pull request merged**